From 71c86c3b73cb9d52297761a59b45832767decea1 Mon Sep 17 00:00:00 2001
From: Icenowy Zheng <icenowy@aosc.io>
Date: Thu, 16 Aug 2018 21:53:02 +0800
Subject: [PATCH 044/139] drm: sun4i: set the sun8i-dw-hdmi TMDS clock
 exclusively when enabling

On H3/H5/A64/R40 the HDMI PHY clock is set after the the mode_set() of
the HDMI encoder is called. As the PHY clock has the same parent with
the TMDS clock (PLL-VIDEO{0,1}), it may mess up TMDS clock.

Set the rate of TMDS clock only when enabling, and add exclusivity to
it. Drop the exclusivity when disabling, to allow the change of other
clocks connected under the PLL.

Signed-off-by: Icenowy Zheng <icenowy@aosc.io>
---
 drivers/gpu/drm/sun4i/sun8i_dw_hdmi.c | 18 +++++++++++++++++-
 drivers/gpu/drm/sun4i/sun8i_dw_hdmi.h |  1 +
 2 files changed, 18 insertions(+), 1 deletion(-)

diff --git a/drivers/gpu/drm/sun4i/sun8i_dw_hdmi.c b/drivers/gpu/drm/sun4i/sun8i_dw_hdmi.c
index 31875b636434..98a715396a8b 100644
--- a/drivers/gpu/drm/sun4i/sun8i_dw_hdmi.c
+++ b/drivers/gpu/drm/sun4i/sun8i_dw_hdmi.c
@@ -20,12 +20,28 @@ static void sun8i_dw_hdmi_encoder_mode_set(struct drm_encoder *encoder,
 {
 	struct sun8i_dw_hdmi *hdmi = encoder_to_sun8i_dw_hdmi(encoder);
 
-	clk_set_rate(hdmi->clk_tmds, mode->crtc_clock * 1000);
+	hdmi->clk_tmds_freq = mode->crtc_clock * 1000;
+}
+
+static void sun8i_dw_hdmi_encoder_enable(struct drm_encoder *encoder)
+{
+	struct sun8i_dw_hdmi *hdmi = encoder_to_sun8i_dw_hdmi(encoder);
+
+	clk_set_rate_exclusive(hdmi->clk_tmds, hdmi->clk_tmds_freq);
+}
+
+static void sun8i_dw_hdmi_encoder_disable(struct drm_encoder *encoder)
+{
+	struct sun8i_dw_hdmi *hdmi = encoder_to_sun8i_dw_hdmi(encoder);
+
+	clk_rate_exclusive_put(hdmi->clk_tmds);
 }
 
 static const struct drm_encoder_helper_funcs
 sun8i_dw_hdmi_encoder_helper_funcs = {
 	.mode_set = sun8i_dw_hdmi_encoder_mode_set,
+	.enable = sun8i_dw_hdmi_encoder_enable,
+	.disable = sun8i_dw_hdmi_encoder_disable,
 };
 
 static const struct drm_encoder_funcs sun8i_dw_hdmi_encoder_funcs = {
diff --git a/drivers/gpu/drm/sun4i/sun8i_dw_hdmi.h b/drivers/gpu/drm/sun4i/sun8i_dw_hdmi.h
index aadbe0a10b0c..251f4f49a69f 100644
--- a/drivers/gpu/drm/sun4i/sun8i_dw_hdmi.h
+++ b/drivers/gpu/drm/sun4i/sun8i_dw_hdmi.h
@@ -171,6 +171,7 @@ struct sun8i_hdmi_phy {
 
 struct sun8i_dw_hdmi {
 	struct clk			*clk_tmds;
+	unsigned long			clk_tmds_freq;
 	struct device			*dev;
 	struct dw_hdmi			*hdmi;
 	struct drm_encoder		encoder;
-- 
2.18.0

